split and rename insight into smaller ones (#2026)
enkidevs/curriculum: python/functional-programming/immutability/atomicity-of-failure.md

      raise ValueError("""You can't have
    less than zero items in the basket!""")
    self.itemcount = itemcount

  def increment_items(self):
    self.itemcount +=1

  def decrement_items(self):
    self.itemcount -=1

  def __repr__(self):
    return("Shopping Basket with " +
     str(self.itemcount) + " items.")
```

Can you see how this constraint could be broken? Let's do it:

```python
b = MutableShoppingBasket(1)
print(b)
# Shopping Basket with 1 items.
b.decrement_items()
print(b)
# Shopping Basket with 0 items.
b.decrement_items()
print(b)
# Shopping Basket with -1 items.
```

This specific `MutableShoppingBasket` is now in a broken state and ready to cause undefined behaviour in other parts of the program and our site. What if we used it in an immutable fashion?

```python
b = MutableShoppingBasket(1)
print(b)
# Shopping Basket with 1 items.
b2 = MutableShoppingBasket(b.itemcount-1)
print(b2)
# Shopping Basket with 0 items.
b3 = MutableShoppingBasket(b2.itemcount-1)
print(b3)
# Traceback (most recent call last):
#   File "python", line 27, in <module>
#   File "python", line 4, in __init__
# ValueError: You can't have less than
#    zero items in the basket!
```

> It doesn't let us, as, in creating the new object, we fail to satisfy the constraints when the constructor is called. The object is not created, and it is easy to spot where we've went wrong due to failure being contained at this point and this point only. (*As opposed to silently creating a broken object to cause trouble elsewhere.*) This is another great advantage of immutable objects as it provides this without much extra effort on the developer's part.

Reviewer: ```suggestion
It won't work. When creating the new object, we fail to satisfy the constraints when the constructor is called. The object is not created, and it is easy to spot where we've went wrong due to failure being contained at this point and this point only (*as opposed to silently creating a broken object to cause trouble elsewhere*). This is another great advantage of immutable objects as it provides this without much extra effort on the developer's part.
```
